aten/src/ATen/native/SpectralOps.cpp

```diff
@@ -5,33 +5,152 @@
 #endif
 
 #include "ATen/ATen.h"
+#include "ATen/Config.h"
 #include "ATen/NativeFunctions.h"
-#include "ATen/WrapDimUtils.h"
-#include "ATen/ExpandUtils.h"
+#include "ATen/native/SpectralOpsUtils.h"
 
 #include <algorithm>
-#include <functional>
-#include <numeric>
 #include <vector>
-
+#include <cmath>
 
 namespace at { namespace native {
 
-// Since real-to-complex satisfy the Hermitian symmetry, i.e.,
-// X[m, \omega] = X[m, N - \omega]*. We return only the first floor(N / 2) + 1
-// values by default. This is also the assumption in libraries including cuFFT.
-static inline int64_t infer_ft_complex_length(int64_t real_length) {
-  return (real_length >> 1) + 1;
+static inline Tensor _fft(const Tensor &self, const int64_t signal_ndim,
```

Comment:
So, is this a pass through function which does argument checking?

Reply:
Yes!

Reply:
Also does some shape inference